examples/ec2_remote/index.ts

```diff
@@ -42,31 +45,38 @@ const connection: types.input.remote.ConnectionArgs = {
 };
 
 const connectionNoDialRetry: types.input.remote.ConnectionArgs = {
-    host: server.publicIp,
-    user: "ec2-user",
-    privateKey: privateKey,
-    dialErrorLimit: -1,
+    ...connection,
+    dialErrorLimit: 1,
 };
 
+// We poll the server until it responds.
+//
+// Because other commands depend on this command, other commands are guaranteed
+// to hit an already booted server.
+const poll = new remote.Command("poll", {
+  connection: { ...connection, dialErrorLimit: -1 },
+    create: "echo 'Connection established'",
+}, { customTimeouts: { create: "10m" } })
```

Comment:
This doesn’t feel like it should be something we require users to do. But it’s also not clear why it helps.  It seems to just do what the code did for the other commands prior to the changes here?

Reply:
> It is required because aws.ec2.Instance returns before the resource is fully up and running. If we want to make this unnecessary the fix will need to be in aws and ultimately in pulumi-aws's upstream.

But prior to these changes, all resources used `dialErrorLimit: -1,` already, right?  So should have been waiting for it to be ready even with no changes?  

It seems the changes to the example moved from (a) all resources use `dialErrorLimit: -1,` to (b) one resources uses `dialErrorLimit: -1,`, and the rest depend on that resource.  I can't tell why that helps.  Why didn't (a) also work for the same reasons (b) does?

Reply:
My concern is - are we saying that to correctly use these resources you have to write code like this?  I hope not - it seems like it is supposed to be the job of `remote.Command` to ensure it can retry for a sufficiently long time to ensure this works in just the "obvious" way?

Reply:
> But prior to these changes, all resources used dialErrorLimit: -1, already, right? So should have been waiting for it to be ready even with no changes?

No. Previously `"hostname"`, `"remotePrivateIP"`, `"size"` and `"checkSize"` (using logical names) all used `connection` to connect. `connection` is defined as:

https://github.com/pulumi/pulumi-command/blob/f0ba2f06be7304bff459cad77bc626d0596529a3/examples/ec2_remote/index.ts#L38-L42

`connection` doesn't specify `dialErrorLimit`, so it defaults to `10`. That's why it doesn't wait forever.

`hostname` was the resource that generally errored. The resources defined with `{ dialErrorLimit: -1 }` did not error.

> My concern is - are we saying that to correctly use these resources you have to write code like this? I hope not - it seems like it is supposed to be the job of remote.Command to ensure it can retry for a sufficiently long time to ensure this works in just the "obvious" way?

It was always necessary to write code like this to remove the race condition. In the past we tended to see the race condition work out in our favor, but it was still there. We would see occasional test failures that were fixed with retry.

We can extend the default retry limit, but if your service takes an hour to become responsive, I don't think this should work by default. If you misconfigure your network so the connection will always be refused, it shouldn't take an hour of retries to fail.

Wait for the server to provision and fail fast are both important and in tension here.